daemon/logger/amqp/amqp.go

```diff
@@ -0,0 +1,419 @@
+		logrus.Errorf("Could not reconnect: %v", err)
+		return err
+	}
+	logrus.Info("Reconnected")
+	s.connection = connection
+	return nil
+}
+
+// Take the log message and publish it to the currently connected broker
+func (s *amqpLogger) Log(msg *logger.Message) (err error) {
+	// Remove trailing and leading whitespace
+	short := bytes.TrimSpace([]byte(msg.Line))
+
+	if s.connection == nil {
+		err = reconnect(s)
+		if err != nil {
+			return err
+		}
+	}
+
+	currentBroker := s.connection.brokerURLs[s.connection.broker]
+
+	// If the message isn't empty then send it to the broker
+	if string(short) != "" {
+		m := amqpMessage{
+			Version:   "1",
+			Host:      s.fields.Hostname,
+			Message:   string(short),
+			Timestamp: time.Now(),
+			Path:      s.fields.ContainerID,
+			Tags:      s.fields,
+		}
+
+		messagejson, err := json.Marshal(m)
+		if err != nil {
+			logrus.Errorf("Could not serialise event - %v", err)
+			return err
+		}
+
+		amqpmsg := amqp.Publishing{
+			DeliveryMode: amqp.Persistent,
+			Timestamp:    time.Now(),
+			ContentType:  "application/json",
+			Body:         messagejson,
+		}
+
+		if currentBroker.Confirm == true && s.connection != nil {
+			defer confirmOne(s.connection.conf)
+		}
+
+		err = s.connection.c.Publish(currentBroker.Exchange, currentBroker.RoutingKey, false, false, amqpmsg)
+		if err != nil {
+			err = reconnect(s)
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+
+	return nil
+}
+
+// If message confirmation is enabled then ensure that the confirmation message is
+// received and dealt with
+func confirmOne(confirms <-chan amqp.Confirmation) {
+	logrus.Debug("Waiting for confirmation of publishing")
+	if confirmed := <-confirms; confirmed.Ack {
+		logrus.Debugf("Confirmed delivery with tag: %d", confirmed.DeliveryTag)
+	} else {
+		logrus.Debugf("Failed delivery with tag: %d", confirmed.DeliveryTag)
+	}
+}
+
+// Take the log-driver options and put them into a data structure. Convert any values
+// to the appropriate data type (eg string -> bool)
+func parseURL(ctx logger.Context) (brokerArray []*amqpBroker) {
+	amqpURLs := strings.Split(ctx.Config["amqp-url"], " ")
+	for _, amqpURL := range amqpURLs {
+		u, err := url.Parse(amqpURL)
+		b, err := strconv.ParseBool(ctx.Config["amqp-confirm"])
+		if err != nil {
+			logrus.Errorf("String %v is not a valid URL: %v", u, err)
+		} else {
+			broker := &amqpBroker{
+				BrokerURL:  u,
+				CertPath:   ctx.Config["amqp-cert"],
+				KeyPath:    ctx.Config["amqp-key"],
+				Exchange:   ctx.Config["amqp-exchange"],
+				Queue:      ctx.Config["amqp-queue"],
+				RoutingKey: ctx.Config["amqp-routingkey"],
+				Tag:        ctx.Config["amqp-tag"],
+				Confirm:    b,
+			}
+			brokerArray = append(brokerArray, broker)
+		}
+	}
+	return brokerArray
+}
+
+// Take the settings specified in the JSON configuration file and put them into a data
+// structure. Convert any values to their required data types
+func parseJSONFile(path string) (brokerArray []*amqpBroker) {
```

Comment:
Yes, this is a complicated file but would only be needed if someone was connecting to a cluster of secure  AMQP brokers. If the user is setting up connections to secure AMQP brokers then they will need to specify the key and certificate files which need to be present on the server. I'm presuming that if someone has gone to the trouble of setting up a cluster of secure AMQP brokers then they will be comfortable with other complicated setups.
If we don't think that the configuration file is going to work then I can see two alternatives. The first is to use the method @runcom suggests, which is also complicated, but would allow anyone to set up connections to clusters, both amqp and amqps. The second is to forget about providing the ability to connect to a cluster and rely on the user to set up a load balancer in front of their cluster. That way the log driver only ever needs to know about one address, regardless of whether it's a cluster or not.